File: python/day07/part2.py
#!/usr/bin/env python3
import time
import part1


def find_min_folder_size(folder, target):
    assert folder.size >= target
    candidates = [folder.size]
    for f in folder.folders:
        if f.size >= target:
            candidates.append(find_min_folder_size(f, target))
    return min(candidates)


def solve_part2(commands):
    root = part1.build_tree(commands)
    target = 30000000 - (70000000 - root.size)
    return find_min_folder_size(root, target)


if __name__ == "__main__":
    start = time.time()
    parsed = part1.parse("data.txt")
    result = solve_part2(parsed)
    print(f'Part 2 : {result}  ({round(1000 * (time.time() - start))} ms)')

File: c++/day07/part2.cpp
#include "part2.h"

using namespace std;


namespace Day07 {
    namespace Part2 {

        void findDirectorySizeToRemove(const shared_ptr<Directory>& directory, int target, int& result) {
            if (directory->size < target) {
                return;
            }
            if (directory->size < result) {
                result = directory->size;
            }
            for (const auto& [_, folder] : directory->folders) {
                findDirectorySizeToRemove(folder, target, result);
            }
        }

        int solve(const shared_ptr<Directory>& directory) {
            const int available = 70'000'000 - directory->size;
            const int target = 30'000'000 - available;

            int result = directory->size;
            findDirectorySizeToRemove(directory, target, result);
            return result;
        }
    }
}
